pkg/selectors/selectors.go

```diff
@@ -13,6 +21,27 @@ type KernelSelectorState struct {
 
 	// valueMaps are used to populate value maps for InMap and NotInMap operators
 	valueMaps []map[[8]byte]struct{}
+
+	// matchBinaries mappings
+	binVals map[uint32]string
+}
+
+func (k *KernelSelectorState) AddBinaryName(binary string) uint32 {
+	if k.binVals == nil {
+		k.binVals = make(map[uint32]string)
+	}
+
+	binMu.Lock()
+	idx := binIdx
+	binIdx++
+	binMu.Unlock()
+
+	k.binVals[idx] = binary
```

Comment:
I think the map assignment needs to also be under the lock.
See: https://go.dev/doc/faq#atomic_maps.

Reply:
Thanks, I have fixed that.

Reply:
hmm this is confusing, am I missing something here? there should be two mutexes one for the global one, and one per kernelselectorstate struct so we don't mix the two?

Reply:
I tried to simplify this a bit. We serialize all calls to `AddBinaryName` even if they refer to different sensors. This is not a performance-critical operation, so I believe this is fine (and simple).

Other than that, I don't believe that there is any reason to have a lock per KernelSelectorState as from my understanding always 1 thread accesses these variables.

Reply:
ok, what I was referring is the global lock binMu is also used to lock the map of KernelSelectorState k.binVals[] per @kkourt comment https://github.com/cilium/tetragon/pull/686/commits/063434c9a04ccc8f61869f3a63e4690141c6e93f#r1104335002  in this patch context and I find it confusing.

So if AddBinaryName() are serialized, then why not have a helper that does the binIdx and binVals update lock/unlock inside, return the appropriate result and hide everything from AddBinaryName()... if that could work.

As said since we want to fix this feel free to merge, and please improve later if it makes sense.

Reply:
> So if AddBinaryName() are serialized

`AddBinaryName` calls are not serialized by design. We serialize them using the global mutex. 

Yes, we could create new functions that do `binIdx` and `binVals` update with their appropriate locking but to be honest I found that a bit overkill. 

We can refactor that in a followup PR to make that even better. Thanks!

Reply:
> > So if AddBinaryName() are serialized
> 
> `AddBinaryName` calls are not serialized by design. We serialize them using the global mutex.
> 

Hence my confusion, why are we using **one** same global mutex to serialize multiple different contexts!

> Yes, we could create new functions that do `binIdx` and `binVals` update with their appropriate locking but to be honest I found that a bit overkill.

There is no way around it, if you grep the code we do it everywhere when updating maps...

> 
> We can refactor that in a followup PR to make that even better. Thanks!

Alright , thanks!